core/src/main/scala/shapeless/ops/hlists.scala

```diff
@@ -2036,20 +2044,24 @@ object hlist {
      */
     override type Aux[P <: HList, S <: HList, Out0 <: HList] = Prepend[P, S] { type Out = Out0 }
 
-    implicit def hnilPrepend0[P <: HList, S <: HNil]: Aux[P, S, P] =
+    implicit def hnilPrepend0[P <: HList, S <: HNil](implicit n: HNil.type <:< S): Aux[P, S, P] =
```

Comment:
Better to just remove the `S` type parameter and use `HNil` (which is also binary compatible).
I wonder if there was a reason to use `S` in the first place.

Reply:
I have given it a try, but `HNil` doesn't seem to cover some cases where `HNil.type` is needed/expected.

```
shapeless/core/src/test/scala/shapeless/hlist.scala:427:53: could not find implicit value for parameter prepend: shapeless.ops.hlist.Prepend[L,shapeless.HNil.type]
```

I think the `S` type parameter is used to cover both `HNil` and `HNil.type` types in one go.
The alternative is writing separate implicits for `HNil` and `HNil.type`, which makes tests pass, but involves some additional code duplication.

Reply:
Ok a quick search for `<: HNil` reveals that it's used 25 times. But definitely not every time which is quite inconsistent. How about doing the ugly thing here and just use `asInstanceOf`?

In hindsight I think it would have been better if didn't expose `object HNil` but just `val HNil: HNil = new HNil {}`. @milessabin WDYT, is this something we ought to do in 2.4?

Reply:
I'm not enthusiastic about using `asInstanceOf` unless there's absolutely no alternative.

> is this something we ought to do in 2.4

I'm in two minds about this. On the one hand I think it'd be good to experiment ... I dimly remember there might have been a good reason for it being the way it is, but maybe it's historical, or maybe it was a bad choice that I wouldn't make now.

On the other, it's been this way since the beginning and I'm sure there will be client code that relies on the more precise type. I'm not sure I want to break source compatibility in that way for 2.4. 2.5 _will_ be breaking souce compatibility, so maybe it'd be better to leave it until then.

Also bear in mind that 3.0 will very likely be deprecating `HList`s in favour of `Unit` and tuples, so it might not be worth changing it at all in the 2.x series.

Reply:
> Also bear in mind that 3.0 will very likely be deprecating HLists in favour of Unit and tuples, so it might not be worth changing it at all in the 2.x series.

Alright, makes sense. Let's not touch it then.

@antbell sorry it took me so long to notice this but we should be able to say `S >: HNil.type <: HNil` right?

Reply:
> @antbell sorry it took me so long to notice this but we should be able to say `S >: HNil.type <: HNil` right?

That works, but I think it's not necessary, since `HNil` is a sealed trait, only implemented by the `HNil` object.

Reply:
Right, I meant instead of the implicit `<:<` evidence

Reply:
That works indeed and is more elegant. Nice! I have pushed the change.